### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by dfreeman**
Initial implementation. I'll open a secondary PR with the dummy app in it, which should give some better examples of component implementations, though there are some basics in `exclaim-ui-test`.

I'll be building out the README later on, so if there's anything in the meantime that doesn't make any sense, give me a shout and I'm happy to talk through it. High level:
 - the way to render an exclaim interface is with the `{{exclaim-ui}}` component, which takes:
   - `ui` - some JSON describing the spec for the interface
   - `env` - an object containing values to be exposed to the rendered components
   - `resolveComponent` - an action that receives the name of a component given in the JSON and returns the full path to the Ember component implementation that should be used
   - `bindKey`, `componentKey` - optional values specifying which keys are "special" in the `ui` configuration (defaults to `$bind` and `$component`)
 - `{{exclaim-ui}}` uses `buildSpecProcessor` to convert the given JSON into a "component spec" for rendering, which it then hands off to `{{exclaim-component}}`
 - `{{exclaim-ui}}` wraps the given `env` object in an `Environment` instance, which handles resolving bound values when requested
 - `Environment` instances have an `extend` method, which returns a new environment that has the given values as well as the ones inherited from its parent
 - component implementations can `{{yield childSpec}}` (or `{{yield childSpec overrideEnv}}`) in order to have `{{exclaim-ui}}` render child components

@deverstalmage 

P.S. almost 1300 lines of this is `yarn.lock`, and another like 300 is tests — the actual implementation isn't too huge.

### CODE REVIEW

File: addon/-private/environment.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import Ember from 'ember';
```
Comment (dfreeman) [inline addon/-private/environment.js]: On reflection, breaking this file apart into a few different modules might help with comprehension (along with some comments) — if you agree, I'll take a pass on that.

Comment (deverstalmage) [inline addon/-private/environment.js]: yeah I think that would help a good amount, taking this in all in one read is pretty tricky when it's all in a vertical line

Comment (dfreeman) [inline addon/-private/environment.js]: Just pushed a change to break this up. I'd recommend starting with `environment/index.js` and `create-env-computed.js`, then moving to `array-value.js` and `value.js`

File: addon/-private/build-spec-processor.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        throw new Error(`Invalid binding: ""`);
+      } else {
+        return new Binding(name);
+      }
+    },
+
+    @rule({ [componentKey]: simple('name'), ...rest('config') })
+    createComponentSpec({ name, config }, { resolveComponent, owner }) {
+      const path = resolveComponent(name);
+      if (!owner.hasRegistration(`component:${path}`)) {
```
Comment (deverstalmage) [inline addon/-private/build-spec-processor.js]: why not pass the component instance and call `getOwner` on it here?

Comment (dfreeman) [inline addon/-private/build-spec-processor.js]: @deverstalmage maybe I'm misunderstanding, but isn't that a chicken-and-egg issue? If I already have a component instance, I don't need to validate that it resolves to something in the container.

If we move to a model where the available components are passed in, I guess this question goes away.

Comment (dfreeman) [inline addon/-private/build-spec-processor.js]: Unless you meant the `{{exclaim-ui}}` component instance, in which case just passing the owner feels better to me from a Law of Demeter perspective

**Comment by deverstalmage** [comment]
shorthand and removal of `bindKey` and `componentKey` looks good :+1:

**Pull request merged**